src/helpers/ConstrainHelpers.ts

```diff
@@ -0,0 +1,207 @@
+  const constrainedModel = new ConstrainedDictionaryModel(context.constrainedName, context.metaModel.originalInput, '', keyModel, valueModel, context.metaModel.serializationType);
+  constrainedModel.type = getTypeFromMapping(typeMapping, {
+    constrainedModel,
+    renderer: context.renderer,
+    propertyKey: context.propertyKey
+  });
+  return constrainedModel;
+}
+
+function constrainObjectModel<R extends AbstractRenderer>(typeMapping: TypeMapping<R>, constrainRules: Constraints, context: ConstrainContext<R, ObjectModel>): ConstrainedObjectModel {
+  const constrainedModel = new ConstrainedObjectModel(context.constrainedName, context.metaModel.originalInput, '', {});
+  for (const [propertyKey, propertyMetaModel] of Object.entries(context.metaModel.properties)) {
+    const constrainedPropertyName = constrainRules.propertyKey({propertyKey, constrainedObjectModel: constrainedModel, objectModel: context.metaModel});
+    const constrainedProperty = constrainMetaModel(typeMapping, constrainRules, {...context, metaModel: propertyMetaModel, propertyKey: constrainedPropertyName});
+    constrainedModel.properties[String(constrainedPropertyName)] = constrainedProperty;
+  }
+  constrainedModel.type = getTypeFromMapping(typeMapping, {
+    constrainedModel,
+    renderer: context.renderer,
+    propertyKey: context.propertyKey
+  });
+  return constrainedModel;
+}
+
+function ConstrainEnumModel<R extends AbstractRenderer>(typeMapping: TypeMapping<R>, constrainRules: Constraints, context: ConstrainContext<R, EnumModel>): ConstrainedEnumModel {
+  const constrainedModel = new ConstrainedEnumModel(context.constrainedName, context.metaModel.originalInput, '', []);
+
+  for (const enumValue of context.metaModel.values) {
+    const constrainedEnumKey = constrainRules.enumKey({enumKey: String(enumValue.key), enumModel: context.metaModel, constrainedEnumModel: constrainedModel});
+    const constrainedEnumValue = constrainRules.enumKey({enumKey: String(enumValue.key), enumModel: context.metaModel, constrainedEnumModel: constrainedModel});
+
+    const constrainedEnumValueModel = new ConstrainedEnumValueModel(constrainedEnumKey, constrainedEnumValue);
+    constrainedModel.values.push(constrainedEnumValueModel);
+  }
+  constrainedModel.type = getTypeFromMapping(typeMapping, {constrainedModel, renderer: context.renderer, propertyKey: context.propertyKey});
+  return constrainedModel;
+}
+
+export function constrainMetaModel<R extends AbstractRenderer>(typeMapping: TypeMapping<R>, constrainRules: Constraints, context: ConstrainContext<R, MetaModel>): ConstrainedMetaModel {
+  const constrainedName = constrainRules.modelName({modelName: context.metaModel.name});
+  const newContext = {...context, constrainedName};
+  if (newContext.metaModel instanceof ObjectModel) {
+    return constrainObjectModel(typeMapping, constrainRules, {...newContext, metaModel: newContext.metaModel});
```

Comment:
You can simplify it:

```ts
const newContext = {...context, constrainedName, metaModel: context.metaModel};

....

return constrainReferenceModel(typeMapping, constrainRules, newContext);
```

Reply:
I tried, TS is throwing errors that the types for metaModel, when done like that, would be `MetaModel`, which does not match with `ReferenceModel` (for example).

Alternatively, I could do:
```
const newContext = {...context, constrainedName};
....
return constrainReferenceModel(typeMapping, constrainRules, newContext as any);
```

But that seems forced as well, and kinda shady 🤔 What you think?

Reply:
Ok, so please don't propagate my suggestion :) LGTM!

Reply:
Alright 👍